### PR TITLE
[GLM-5.3-Flash] Rework kpool compress and dual-stream indexer scheduling

### PR DESCRIPTION
**Pull request by Dovis01**
## Purpose

Supplements vLLM's `Glm5Next` sparse-MLA kpool indexer with the kpool optimizations

What changes in vLLM:

- `vllm/models/glm5next/{nvidia,amd}/ops/kpool_compress.py`: `_kpool_softmax_rotate_write_cache_kernel` switches from a two-pass to an **online softmax** (each pool's gate scores are read once instead of twice) and takes a (row stride, slot stride, row offset) triple so one kernel serves both the pooled `[n_pools, kpool, D]` layout (`row_offset = 0`) and the **flat token layout** (`row_offset = kpool - 1`: row `i` pools tokens `i - kpool + 1 .. i`, SGLang's "assemble" kernel shape). Rows whose slot is negative, or too early in the batch to own a full pool, return before any load — replacing the `write_mask` mechanism. The prefill and decode kernels share one `_hadamard_quantize_fp8` helper; the decode/verify kernel also uses the online softmax (matching SGLang's verify-side kernel).
- `vllm/models/glm5next/{nvidia,amd}/sparse_indexer.py`: the prefill insert becomes a single launch of the new `kpool_compress_tokens_and_write_cache` on the prefill slice (was: `arange`/`clamp`/`>=`/`&`/index tensors + `k[idx]`/`gate_score[idx]` gathers + int64 conversion = 12 launches per indexer layer per prefill step). On **eager pure-prefill batches** the compress + tail-seed writes now run on the global aux stream, overlapping the current stream's topk-buffer fill and gather/logits preparation, joined before the first KV gather (SGLang's `_can_overlap_prefill`). The overlap is gated off under CUDA graph capture and breakable CUDA graphs.
- `vllm/compilation/breakable_cudagraph.py`: new `is_in_breakable_cuda_graph()` (true during breakable capture or replay) so cross-stream code can stay single-streamed under breakable graphs — the vLLM counterpart of SGLang's guard with the same rationale (the BCG eager break must finish its indexer work on one stream).
- `vllm/models/glm5next/common/attention.py` (`Indexer.forward`): the head gate is `torch.mm(hidden, w.t(), out_dtype=torch.float32)` on the bf16 operands on CUDA (fp32 accumulation and output, no `hidden.float()` copy, no cached fp32 weight; the fp32-upcast fallback is kept for other platforms). Under CUDA graph capture the K path (wk GEMM, k_norm, compress-gate GEMM) runs on the aux stream while the current stream projects, Hadamard-rotates and quantizes Q and computes the head gate — the two streams join before the indexer op (SGLang's "rebalance dual-stream indexer preparation"). Eager batches stay sequential (the fork/join would cost host time there).

## Duplicate-work check

`kpool`, `GLM-5.3-Flash indexer` and `glm5next` (2026-09-15) shows no PR touching the compress kernels, the prefill insert path, the head-gate GEMM or the indexer's stream scheduling. Related but different open PRs: #55358 moved `sparse_attn_indexer_kpool.py` into the model folder (a refactor this PR is rebased on top of), #54951 shards long-context indexer prefill rows across TP, #56607 / #56059 are ROCm fixes to the Q quant and the tail-seed kernel, #55270 fixes the paged-MQA varlen launch mode.

## Perf

Kernel-level microbenchmarks (H100, GLM-5.3-Flash shapes: kpool 4, head_dim 128, 32 indexer heads, page 64):

- Prefill compress-write path per indexer layer: 12 launches -> 1. Wall time 154-216 us -> 32-36 us at n = 256 / 2048 / 8192 tokens (-79% .. -84%); summed GPU kernel time 29.0 -> 4.4 us at n = 2048 and 67.9 -> 6.7 us at n = 8192. Output is byte-identical to the old path except 9 of 278,784 cache bytes at n = 8192 (online vs two-pass softmax rounding in fp32, then bf16 -> fp8).
- `Indexer.forward` GEMM/quant sequence, CUDA-graph replay time: T = 48 tokens 39.7 -> 34.3 us (head gate) -> 22.7 us (head gate + dual stream), -43%; T = 288 (48 sequences x MTP 6) 50.8 -> 30.8 us, -39%; T = 1024: 94.4 -> 54.3 us, -42%; T = 4096 (head gate only; the stream overlap is off above 1024 tokens): 255.4 -> 172.7 us, -32%. Head-gate weights differ by at most 1.3e-9 (fp32 summation order).
- Decode tail-update kernel: unchanged time (launch-bound, ~36.5 us at B = 48, next_n 1 and 6) and byte-identical output.

End-to-end (`vllm bench serve`, GLM-5.3-Flash, 8xH100, `--tensor-parallel-size 8 --enable-expert-parallel --max-num-seqs 48 --max-model-len 131072 --speculative-config '{"method":"mtp","num_speculative_tokens":5}'`, breakable CUDA graphs; client `--dataset-name random --ignore-eos --seed 0`; main = e0301afb2a, same machine, back-to-back runs):

| Config/conc 1 | Metric | main | this PR | Delta |
|---|---|---|---|---|
| 8192 in / 16 out, 32 prompts | mean TTFT | 440.2 ms | 426.3 ms | -3.2% |
| | median TTFT | 432.5 ms | 422.4 ms | -2.3% |
| | p99 TTFT | 513.5 ms | 502.8 ms | -2.1% |
| | mean E2EL | 683.4 ms | 665.5 ms | -2.6% |
| | output throughput | 23.4 tok/s | 24.0 tok/s | +2.7% |
| 2048 in / 256 out, 96 prompts | mean TTFT | 390.8 ms | 384.4 ms | -1.6% |
| | mean ITL | 18.13 ms | 18.05 ms | -0.4% |
| | mean TPOT | 11.55 ms | 11.65 ms | +0.9% |
| | output throughput | 76.8 tok/s | 76.3 tok/s | -0.6% |

The prefill-side changes land where they work: TTFT drops 2-3% on prefill-heavy traffic (11 indexer layers x 11 fewer launches, ~60 us less GPU time and one overlapped write phase per layer per prefill chunk). Decode-heavy traffic is neutral within run-to-run noise (the decode-side kernel is launch-bound and unchanged).

Higher-concurrency runs (same servers and flags as the table above):

| Config | Metric | main | this PR | Delta |
|---|---|---|---|---|
| 8192 in / 16 out, 128 prompts, conc 16 | mean TTFT | 1321.5 ms | 1359.7 ms | +2.9% |
| | median TTFT | 1210.8 ms | 1220.6 ms | +0.8% |
| | mean E2EL | 4583.6 ms | 4364.1 ms | -4.8% |
| | output throughput | 55.3 tok/s | 58.1 tok/s | +5.1% |
| 2048 in / 256 out, 256 prompts, conc 32 | mean TTFT | 805.6 ms | 801.7 ms | -0.5% |
| | p99 TTFT | 2855.6 ms | 2823.6 ms | -1.1% |
| | mean ITL | 96.98 ms | 99.25 ms | +2.3% |
| | output throughput | 629.8 tok/s | 621.2 tok/s | -1.4% |

At concurrency the prefill-heavy config gains 5.1% output throughput and 4.8% mean E2EL while individual TTFTs (queueing-dominated, chunked into mixed batches) stay within a few percent; the decode-heavy config is neutral within run-to-run noise.

## Tests

- `tests/kernels/test_kpool_decode_update_batched.py`: 3 new tests — the flat-token writer equals the pooled writer on explicitly gathered rows across a request boundary that is not pool-aligned inside the batch (kpool 4 and 16), a slot on a row too early to own a full pool is skipped, and a batch shorter than one pool writes nothing. The existing exact-equality tests against the two-pass fp32 CPU reference (incl. 20 fuzz seeds) and the decode-vs-prefill writer equality keep passing with the online-softmax kernels. Whole file: 31 passed, 2 skipped (ROCm-only). `tests/v1/attention/test_kpool_tail_slot_mapping.py`, `tests/v1/attention/test_sparse_indexer_decode_seq_lens.py`: pass (50 passed / 2 skipped combined).
- Live server (breakable CUDA graphs: prefill PIECEWISE/FULL + decode FULL graphs captured cleanly with the dual-stream fork/join inside, MTP speculator): greedy sanity (`17 * 23` -> `391` with a correct reasoning trace, identical to main) and the bench runs above.
- GSM8K (`tests/evals/gsm8k/gsm8k_eval.py --num-questions 300 --max-concurrency 1`, 5-shot, greedy, MTP k=5, same servers): this PR 0.920, main 0.923 on the paired run; across all runs on this machine both sides span 0.920-0.923 (greedy outputs are not bitwise reproducible across runs on this fp8 MoE + EP deployment even on unmodified main), so this is parity, not an accuracy claim. The only numerics that changed are bounded: 9 of 278,784 cache bytes on an 8192-token prefill (fp8 rounding from the online softmax) and <= 1.3e-9 fp32 summation order in the head gate.
- Pre-commit hooks (ruff, ruff-format, mypy, typos, ...) pass on all changed files.

### CODE REVIEW

File: vllm/models/glm5next/amd/sparse_indexer.py [hunk 1]
```diff
@@ -189,38 +158,61 @@ def sparse_attn_indexer_kpool(
...
+                # writes on the aux stream, overlapping the current stream's
+                # topk-buffer fill and gather/logits preparation (SGLang's
+                # _can_overlap_prefill). Runs under CUDA graph capture or
+                # breakable graphs stay single-streamed.
+                if (
+                    not has_decode
+                    and current_platform.is_cuda()
+                    and not _capturing_cudagraph()
+                    and not is_in_breakable_cuda_graph()
+                ):
```
Comment (JaredforReal) [inline vllm/models/glm5next/amd/sparse_indexer.py]: double check if we need the same change in amd side
`current_platform.is_cuda()` is making this block a dead code

**Review by JaredforReal — COMMENTED** [review]
before look into detail,can we double check if amd implementation need this?

File: vllm/models/glm5next/amd/sparse_indexer.py (hunk 1 above)
Comment (Dovis01) [reply vllm/models/glm5next/amd/sparse_indexer.py]: Yeah. Thx! Done for clearing. Sry.

File: vllm/models/glm5next/amd/ops/kpool_compress.py [hunk 2]
```diff
@@ -180,17 +169,82 @@ def _kpool_softmax_rotate_write_cache_kernel(
...
+    ape = ape.contiguous()
+    loc = loc.contiguous()
+    if return_compressed:
+        compressed_k = torch.empty((n_rows, head_dim), dtype=FP8_DTYPE, device=k.device)
+        compressed_scale = torch.empty((n_rows,), dtype=torch.float32, device=k.device)
+    else:
+        compressed_k, compressed_scale = buf_fp8, buf_fp32
+
+    if n_rows > 0:
+        _kpool_softmax_rotate_write_cache_kernel[(n_rows,)](
```
Comment (JaredforReal) [inline vllm/models/glm5next/amd/ops/kpool_compress.py]: we need unit test for amd's online softmax
